### PR TITLE
[GEOS-11087] Fix IsolatedCatalogFacade unnecessary performance overhead

### PR DESCRIPTION
**Pull request by groldan**
[GEOS-11087](https://osgeo-org.atlassian.net/browse/GEOS-11087)

* Filter the iterator created by list(...) directly instead of building an ArrayList first
* Apply in-process filter to iterator in count() only if required, delegate to the proxied facade's count() method if `Dispatcher.REQUEST.get() == null`

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/catalog/impl/IsolatedCatalogFacade.java [hunk 1]
```diff
@@ -670,15 +673,19 @@ private boolean canSeeWorkspace(WorkspaceInfo workspace) {
...
-                unwrapped.stream()
-                        .filter(store -> filter.apply(store) != null)
-                        .collect(Collectors.toList()),
-                type);
+        if (isOwsRequest()) {
+            // unwrap the catalog objects list
+            List<T> unwrapped = ModificationProxy.unwrap(objects);
+            // filter the non visible catalog objects and wrap the resulting list with a
+            // modification
+            // proxy
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/impl/IsolatedCatalogFacade.java:681]:
```suggestion
            // modification proxy
```

**Review by aaime — COMMENTED** [review]
Looks good to me. Please play nice with the automated formatter :rofl:

File: src/main/src/main/java/org/geoserver/catalog/impl/IsolatedCatalogFacade.java (hunk 1 above)
Comment (groldan) [reply src/main/src/main/java/org/geoserver/catalog/impl/IsolatedCatalogFacade.java]: oops. Fixed.

**Review by aaime — APPROVED** [review]
Yep, looks good now, thanks!

**Pull request merged**